storage/jellyfish_merkle/src/node_type/mod.rs

```diff
@@ -0,0 +1,513 @@
+        }
+    }
+
+    /// A shortcut to generate a node key consisting of a version and empty nibble path.
+    pub fn new_empty_path(version: Version) -> Self {
+        Self::new(version, NibblePath::new(vec![]))
+    }
+
+    /// Gets the version.
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    /// Gets the nibble path.
+    pub fn nibble_path(&self) -> &NibblePath {
+        &self.nibble_path
+    }
+
+    /// Generates a child node_key based on this node key.
+    pub fn gen_child_node_key(&self, version: Version, n: u8) -> Self {
+        let mut node_nibble_path = self.nibble_path().clone();
+        node_nibble_path.push(n);
+        Self::new(version, node_nibble_path)
+    }
+
+    /// Generates a new node key with the same nibble path but a different version.
+    pub fn update_version(&self, version: Version) -> Self {
+        Self {
+            version,
+            nibble_path: self.nibble_path.clone(),
+        }
+    }
+}
+
+/// Each child of [`InternalNode`] encapsulates a nibble forking at this node.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct Child {
+    // The hash value of this child node.
+    pub hash: HashValue,
+    // `version`, the `nibble_path` of the [`InternalNode`] the child belongs to and the child's
+    // index constitute the [`NodeKey`] to uniquely identify this child node from the storage.
+    pub version: Version,
+    // Whether the child is a leaf node.
+    pub is_leaf: bool,
+}
+
+impl Child {
+    pub fn new(hash: HashValue, version: Version, is_leaf: bool) -> Self {
+        Self {
+            hash,
+            version,
+            is_leaf,
+        }
+    }
+}
+
+/// [`Children`] is just a collection of children belonging to a [`InternalNode`], indexed from 0 to
+/// 15, inclusive.
+pub(crate) type Children = HashMap<u8, Child>;
+
+/// Represents a 4-level subtree with 16 children at the bottom level. Theoretically, this reduces
+/// IOPS to query a tree by 4x since we compress 4 levels in a standard Merkle tree into 1 node.
+/// Though we choose the same internal node structure as that of Patricia Merkle tree, the root hash
+/// computation logic is similar to a 4-level sparse Merkle tree except for some customizations. See
+/// the `CryptoHash` trait implementation below for details.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct InternalNode {
+    // Up to 16 children.
+    children: Children,
+}
+
+/// Represents an account. It has two fields: `key` is the hash of the acccont address and
+/// `value_hash` is the hash of account state blob.
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct LeafNode {
+    // The hashed account address associated with this leaf node.
+    account_key: HashValue,
+    // the hash of the account state blob.
+    blob_hash: HashValue,
+    // The account blob associated with `account_key`.
+    blob: AccountStateBlob,
+}
+
+#[repr(u8)]
```

Comment:
This doesn't seem useful?

Reply:
why? I think it define the layout in momery.

Reply:
Does the layout in memory matter? Basically the `derive` is all you use.